File: client/src/app/application/decks/delete-deck.use-case.ts
import { Injectable, inject } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { DeckRepository } from '../../domain/ports/deck.repository';

/** Deletes a deck with its cards and study history. */
@Injectable({ providedIn: 'root' })
export class DeleteDeckUseCase {
  private readonly decks = inject(DeckRepository);

  execute(deckId: string): Promise<void> {
    return firstValueFrom(this.decks.remove(deckId));
  }
}

File: server/LingoHop.Application/Decks/UseCases/DeleteDeckUseCase.cs
using LingoHop.Application.Abstractions;
using LingoHop.Application.Common;
using LingoHop.Domain.Decks;

namespace LingoHop.Application.Decks.UseCases;

/// <summary>Deletes a deck together with its cards and study history.</summary>
public sealed class DeleteDeckUseCase(
    ICurrentLearner currentLearner,
    IDeckRepository decks,
    IUnitOfWork unitOfWork)
{
    public async Task<Result> ExecuteAsync(Guid deckId, CancellationToken cancellationToken = default)
    {
        var user = await currentLearner.GetAsync(cancellationToken);
        var deck = await decks.FindByIdAsync(deckId, cancellationToken);
        if (deck is null || !deck.IsOwnedBy(user.Id))
        {
            return Result.Failure(DeckErrors.NotFound(deckId));
        }

        decks.Remove(deck);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return Result.Success();
    }
}
